qmail-1.03/dk-filter.8
.TH dk-filter 8
.SH NAME
dk-filter \- sign/verify using DK/DKIM (SSP/ADSP optionally) and deliver a mail message for delivery
.SH SYNOPSIS
.B FILTERARGS=/var/indimail/bin/dk-filter
.SH DESCRIPTION
.B dk-filter
is a qfilter which can be set as a filter for
.BR spawn-filter(8) .
The filter can be set either as
.B FILTERARGS
or in the control file
.BR filterargs .

.B dk-filter
supports DK/DKIM signing and verification and can optionally use
.B Sender Signing Practice (SSP)
or
.B Author Domain Signing Practice.
It uses the libdkim and OpenSSL libraries.  To sign a message, set the
.B DKIMSIGN
or
.B DKSIGIN
environment variables to the pathname of the private key that will be
used to sign the message. If there is a % character in the environment
variable, it is removed and replaced by the domain name in the From: header.
If, after substituting the %, that file does not exist, /var/indimail/control/domainkeys/%/default
will be used as the key. If again, after substituting the % sign, the file does not exist,
/var/indimail/control/domainkeys/default will be used as the key.
After all substitutions, if the key file does not exist, the message will not be signed.
If there is no % and the file does not exist, the message will be rejected with error 35.
The selector (s=) will be taken from the basename of the file.
The private key should be created by 
.BR dknewkey(8) .

You can set various DK options in getopt style, by setting the environment variable DKSIGNOPTIONS
 b <advice_length>    Length of Advice
 c <canonicalization> simple, nofws
 s <privkeyfile>
 h                    show headers included
 r                    Skip Duplicate Headers
.EX
 DKSIGNOPTIONS="-h -r -c nofws"
 sets the h= tag, skips duplicate headers and sets nofws canonicalization
.EE

You can set various DKIM options in getopt style, by setting the environment variable DKIMSIGNOPTIONS

 b <standard>         1 - allman, 2 - ietf or 3 - both
 c <canonicalization> r for relaxed [DEFAULT], s - simple,
                      t relaxed/simple, u - simple/relaxed
 l                    include body length tag
 q                    include query method tag;
 t                    include a timestamp tag
 h
 i <identity>         the identity, if not provided it will not be included
 x <expire_time>      the expire time in seconds since epoch
                      ( DEFAULT = current time + 604800)
                      if set to - then it will not be included
 z <hash>             1 for sha1, 2 for sha256, 3 for both
 s <privkeyfile>
 y <selector>

.EX
 DKIMSIGNOPTIONS="-b 1 -c r -q"
 sets allman standard, with relaxed canonicalization and include query method tag
.EE

.B dk-filter
uses the domain found in the Sender: header to set the domain tag. If not it uses the From: header. You can override this by
setting 
.B DKIMDOMAIN
environment variable.
.B DKIMDOMAIN
can be set to an email address or a domain (without the at sign).
To verify a message, set the
.B DKIMVERIFY
or
.B DKVERIFY
environment variables
.B dk-filter
always inserts the 
.B DKIM-Status
or
.B DomainKey-Status
header, so that messages can be
rejected later at delivery time, or in the mail reader. In that case you may set
.B DKIMVERIFY
or
.B DKVERIFY
to an empty string.
.B dk-filter
does not use any signing practice byd default. You can override this by setting the SIGN_PRACTICE to ssp or adsp (lowercase).

.SH "EXIT CODES"
0 for success, non-zero failure

.SH "SEE ALSO"
dknewkey(8),
dktest(8),
dkim(8),
spawn-filter(8)


qmail-1.03/dk-filter.sh
#
# $Log: dk-filter.sh,v $
# Revision 1.16  2013-08-17 15:59:21+05:30  Cprogrammer
# do not treat duplicate DomainKey-Signature as an error
#
# Revision 1.15  2013-08-17 15:02:06+05:30  Cprogrammer
# fixed syntax errors and private key lookup
#
# Revision 1.14  2011-02-10 22:47:01+05:30  Cprogrammer
# fixed exit code of dk-filter when doing verification
#
# Revision 1.13  2011-02-08 22:02:29+05:30  Cprogrammer
# use sender domain when replacing '%' in private key
#
# Revision 1.12  2010-05-04 08:37:42+05:30  Cprogrammer
# do DK signing before DKIM signing to prevent DK_SYNTAX error
#
# Revision 1.11  2009-12-10 19:25:13+05:30  Cprogrammer
# added RCS id
#
# Revision 1.10  2009-12-10 16:41:14+05:30  Cprogrammer
# continue of message gives DK_SYNTAX_ERR
#
# Revision 1.9  2009-05-04 10:30:32+05:30  Cprogrammer
# fixed argument expected error
#
# Revision 1.8  2009-04-21 20:42:44+05:30  Cprogrammer
# added check for dktest, dkim executables
#
# Revision 1.7  2009-04-20 10:06:58+05:30  Cprogrammer
# added DKSIGNOPTS
#
# Revision 1.6  2009-04-19 13:38:24+05:30  Cprogrammer
# added full set of dkim options
# replaced indimail/bin/echo with echo 1>&2
#
# Revision 1.5  2009-04-06 16:37:50+05:30  Cprogrammer
# added SIGN_PRACTICE
# use ietf standard insted of allman so that Yahoo verification does not fail
#
# Revision 1.4  2009-04-03 14:39:00+05:30  Cprogrammer
# added return status
#
# Revision 1.3  2009-04-03 08:55:29+05:30  Cprogrammer
# print error messages to stderr
#
# Revision 1.2  2009-04-02 20:36:25+05:30  Cprogrammer
# added -h option to dktest
# added -x - option to dkim
#
# Revision 1.1  2009-04-02 14:52:27+05:30  Cprogrammer
# Initial revision
#
# $Id: dk-filter.sh,v 1.16 2013-08-17 15:59:21+05:30 Cprogrammer Exp mbhangui $
#
if [ -z "$QMAILREMOTE" -a -z "$QMAILLOCAL" ]; then
	echo "dk-filter should be run by spawn-filter" 1>&2
	exit 1
fi
dksign=0
dkimsign=0
dkverify=0
dkimverify=0
if [ -z "$DKSIGN" -a -z "$DKVERIFY" ] ; then
	DKSIGN=QMAILHOME/control/domainkeys/%/default
	dksign=2
fi
if [ -z "$DKIMSIGN" -a -z "$DKIMVERIFY" ] ; then
	DKIMSIGN=QMAILHOME/control/domainkeys/%/default
	dkimsign=2
fi
if [ ! -z "$DKSIGN" ] ; then
	if [ ! -f QMAILHOME/bin/dktest ] ; then
		echo "QMAILHOME/bin/dktest: No such file or directory" 1>&2
		exit 1
	fi
	percent_found=0
	echo $DKSIGN|grep "%" >/dev/null 2>&1
	if [ $? -eq 0 ] ; then
		percent_found=1
	fi
	if [ ! " $_SENDER" = " " ] ; then
		# replace '%' in filename with domain
		domain=`echo $_SENDER | cut -d@ -f2`
		dkkeyfn=`echo $DKSIGN | sed s{%{$domain{g`
	else
		dkkeyfn=$DKSIGN
	fi
	if [ $dksign -eq 2 -a ! -f $dkkeyfn ] ; then
		dkkeyfn=QMAILHOME/control/domainkeys/default
	fi
	if [ -f $dkkeyfn ] ; then
		dksign=1
	else
		dksign=0
	fi
	if [ $dksign -eq 0 -a $percent_found -ne 1 ] ; then
		exit 35
	fi
	dkselector=`basename $dkkeyfn`
fi
if [ ! -z "$DKIMSIGN" ] ; then
	if [ ! -f QMAILHOME/bin/dkim ] ; then
		echo "QMAILHOME/bin/dkim: No such file or directory" 1>&2
		exit 1
	fi
	percent_found=0
	echo $DKIMSIGN|grep "%" >/dev/null 2>&1
	if [ $? -eq 0 ] ; then
		percent_found=1
	fi
	if [ ! " $_SENDER" = " " ] ; then
		# replace '%' in filename with domain
		domain=`echo $_SENDER | cut -d@ -f2`
		dkimkeyfn=`echo $DKIMSIGN | sed s{%{$domain{g`
	else
		dkimkeyfn=$DKIMSIGN
	fi
	if [ $dkimsign -eq 2 -a ! -f $dkimkeyfn ] ; then
		dkimkeyfn=QMAILHOME/control/domainkeys/default
	fi
	if [ -f $dkimkeyfn ] ; then
		dkimsign=1
	else
		dkimsign=0
	fi
	if [ $dkimsign -eq 0 -a $percent_found -ne 1 ] ; then
		exit 35
	fi
	dkimselector=`basename $dkimkeyfn`
fi
if [ ! -z "$DKVERIFY" ] ; then
	if [ ! -f QMAILHOME/bin/dktest ] ; then
		echo "QMAILHOME/bin/dktest: No such file or directory" 1>&2
		exit 1
	fi
	dkverify=1
fi
if [ ! -z "$DKIMVERIFY" ] ; then
	if [ ! -f QMAILHOME/bin/dkim ] ; then
		echo "QMAILHOME/bin/dkim: No such file or directory" 1>&2
		exit 1
	fi
	dkimverify=1
fi
cat > /tmp/dk.$$
if [ $dkimsign -eq 1 ] ; then
	# DKIMSIGNOPTIONS="-z 1 -b 2 -x - -y $dkimselector -s $dkimkeyfn"
	set -- `getopt lqthb:c:d:i:x:z:y:s: $DKIMSIGNOPTIONS`
	bopt=0
	xopt=0
	zopt=0
	yopt=0
	sopt=0
	dkimopts="QMAILHOME/bin/dkim"
	while [ $1 != -- ]
	do
		case $1 in
		-l)
		dkimopts="$dkimopts -l"
		;;
		-q)
		dkimopts="$dkimopts -q"
		;;
		-t)
		dkimopts="$dkimopts -t"
		;;
		-h)
		dkimopts="$dkimopts -h"
		;;

		-b)
		bopt=1
		dkimopts="$dkimopts -b $2"
		shift
		;;

		-c)
		dkimopts="$dkimopts -c $2"
		shift
		;;

		-i)
		dkimopts="$dkimopts -i $2"
		shift
		;;

		-x)
		xopt=1
		dkimopts="$dkimopts -x $2"
		shift
		;;

		-z)
		zopt=1
		dkimopts="$dkimopts -z $2"
		shift
		;;

		-y)
		yopt=1
		dkimopts="$dkimopts -y $2"
		shift
		;;

		-s)
		sopt=1
		dkimopts="$dkimopts -s $2"
		shift
		;;
		esac
		shift   # next flag
	done
	if [ $zopt -eq 0 ] ; then
		dkimopts="$dkimopts -z 1"
	fi
	if [ $bopt -eq 0 ] ; then
		dkimopts="$dkimopts -b 2"
	fi
	if [ $xopt -eq 0 ] ; then
		dkimopts="$dkimopts -x -"
	fi
	if [ $yopt -eq 0 ] ; then
		dkimopts="$dkimopts -y $dkimselector"
	fi
	if [ $sopt -eq 0 ] ; then
		dkimopts="$dkimopts -s $dkimkeyfn"
	fi
	exec 0</tmp/dk.$$
	eval $dkimopts
	if [ $? -ne 0 ] ; then
		/bin/rm -f /tmp/dk.$$
		exit 1
	fi
fi
if [ $dksign -eq 1 ] ; then
	#dktest: [-f] [-b advice_length] [-c nofws|simple] [-v|-s selector] [-h] [-t#] [-r] [-T][-d dnsrecord]
	# DKSIGNOPTIONS="-z 1 -b 2 -x - -y $dkimselector -s $dkimkeyfn"
	set -- `getopt hrb:c:s: $DKSIGNOPTIONS`
	dkopts="QMAILHOME/bin/dktest"
	sopt=0
	while [ $1 != -- ]
	do
		case $1 in
		-h)
		dkopts="$dkopts -h"
		;;
		-r)
		dkopts="$dkopts -r"
		;;

		-b)
		dkopts="$dkopts -b $2"
		shift
		;;

		-c)
		dkopts="$dkopts -c $2"
		shift
		;;

		-s)
		sopt=1
		dkopts="$dkopts -s $2"
		shift
		;;
		esac
		shift   # next flag
	done
	if [ $sopt -eq 0 ] ; then
		dkopts="$dkopts -s $dkkeyfn"
	fi
	exec 0</tmp/dk.$$
	#QMAILHOME/bin/dktest -h -s $dkkeyfn
	eval $dkopts
	exit_val=$?
	# allow error due to duplicate DomainKey-Header
	if [ $exit_val -ne 0 -a $exit_val -ne 12 ] ; then
		/bin/rm -f /tmp/dk.$$
		exit $exit_val
	fi
fi
if [ $dkimverify -eq 1 ] ; then
	practice=$SIGN_PRACTICE
	if [ " $practice" = " " ] ; then
		practice=0
	elif [ " $practice" = " ssp" ] ; then
		practice=1
	elif [ " $practice" = " adsp" ] ; then
		practice=2
	fi
	exec 0</tmp/dk.$$
	QMAILHOME/bin/dkim -p $practice -v
	ret=$?
	case $ret in
		14)
		/bin/rm -f /tmp/dk.$$
		exit 100
		;;
		88)
		/bin/rm -f /tmp/dk.$$
		exit 111
		;;
	esac
	if [ $ret -lt 0 ] ; then
		/bin/rm -f /tmp/dk.$$
		exit 1
	fi
fi
if [ $dkverify -eq 1 ] ; then
	exec 0</tmp/dk.$$
	QMAILHOME/bin/dktest -v
	if [ $? -ne 0 ] ; then
		/bin/rm -f /tmp/dk.$$
		exit 1
	fi
fi
exec 0</tmp/dk.$$
/bin/rm -f /tmp/dk.$$
cat
exit $?
